### PR TITLE
Projection layer at front of decoders and classifier

### PR DESCRIPTION
**Pull request by 51N84D**
The pretrained encoder gives us a very large latent space (dim=2048), which results in way too many parameters in the decoders and feature-classifier (even without resblocks). To solve this we can learn a projection at each decoder to reduce the dimensionality.

### CODE REVIEW

File: omnigan/blocks.py [hunk 1]
```diff
@@ -239,6 +239,7 @@ def __init__(
         self,
         n_upsample=4,
         n_res=4,
+        input_dim=2048,
         dim=64,
```
Comment (vict0rsch) [inline omnigan/blocks.py:243]: can we rename that to something like `projection_dim` or somehting? to make it extra clear?

File: omnigan/classifier.py [hunk 2]
```diff
@@ -16,22 +17,23 @@ def get_classifier(opts, latent_space, verbose):
 
 
 class OmniClassifier(nn.Module):
-    def __init__(self, latent_space, loss):
+    def __init__(self, latent_space, dim, loss):
```
Comment (vict0rsch) [inline omnigan/classifier.py:20]: sam here

File: omnigan/losses.py [hunk 3]
```diff
@@ -270,17 +273,15 @@ def get_losses(opts, verbose):
...
     # ----------------------------------
     # -----  Discriminator Losses  -----
     # ----------------------------------
     losses["D"] = GANLoss(
-        soft_shift=opts.dis.soft_shift,
-        flip_prob=opts.dis.flip_prob,
-        verbose=verbose,
-    )   
+        soft_shift=opts.dis.soft_shift, flip_prob=opts.dis.flip_prob, verbose=verbose,
+    )
```
Comment (vict0rsch) [inline omnigan/losses.py:286]: @alexrey88 can you set up flake8 and black (with max line length 88)? if <- does not make sense let's chat on slack

File: shared/trainer/defaults.yaml [hunk 4]
```diff
@@ -148,6 +148,7 @@ classifier:
   dropout: 0.4 # probability of being set to 0
   init_type: kaiming
   init_gain: 0.2
+  dim: 128
```
Comment (vict0rsch) [inline shared/trainer/defaults.yaml:151]: renaming + comment to describe this param's role

File: shared/trainer/defaults.yaml [hunk 5]
```diff
@@ -57,7 +57,7 @@ gen:
     n_downsample: &n_downsample 3 # number of downsampling layers in encoder | dim 32 + down 3 => z = 256 x 32 x 32
     n_upsample: *n_downsample # upsampling in spade decoder ; should match encoder.n_downsample
     pad_type: reflect # padding type [zero/reflect]
-    res_dim: 256 # Resblock number of channels (=latent space's), should be 2048 if using deeplabv2
+    res_dim: 2048 # Resblock number of channels (=latent space's), should be 2048 if using deeplabv2
     res_norm: instance # ResBlock normalization ; one of {"batch", "instance", "layer", "adain", "spectral", "none"}
```
Comment (vict0rsch) [inline shared/trainer/defaults.yaml:61]: can you also change this to "spectral" ?

File: omnigan/losses.py (hunk 3 above)
Comment (alexrey88) [reply omnigan/losses.py:286]: I already have both, with --line-length==88 for black, and when I hit save it divides the line in 3.. :/

File: omnigan/classifier.py [hunk 6]
```diff
@@ -16,22 +17,24 @@ def get_classifier(opts, latent_space, verbose):
...
 
 class OmniClassifier(nn.Module):
-    def __init__(self, latent_space, loss):
+    def __init__(self, latent_space, proj_dim, loss):
         super(OmniClassifier, self).__init__()
         assert len(latent_space) == 3
         self.channels = latent_space[0]
         self.feature_size = latent_space[1]
         self.loss = loss
+        dim = proj_dim
```
Comment (vict0rsch) [inline omnigan/classifier.py:26]: are you lazy or do you have a good reason 🛩️

Comment (vict0rsch) [reply omnigan/classifier.py:26]: you know `cmd+D` will select next occurence and you can replace easily :p

File: omnigan/generator.py [hunk 7]
```diff
@@ -228,8 +229,8 @@ def __init__(self, opts):
         super().__init__(
             opts.gen.s.n_upsample,
             opts.gen.s.n_res,
-            opts.gen.s.res_dim,
-            opts.gen.s.output_dim,
+            dim=opts.gen.s.res_dim,
```
Comment (vict0rsch) [inline omnigan/generator.py:232]: `proj_dim`?

Comment (51N84D) [reply omnigan/generator.py:232]: Haha I would eventually have changed that once I started implementing the segmentation head

File: omnigan/classifier.py (hunk 6 above)
Comment (51N84D) [reply omnigan/classifier.py]: You already know the answer, don't you...

**Comment by vict0rsch** [comment]
2020-06-09 at 16 34
very nice commit

File: omnigan/classifier.py (hunk 6 above)
Comment (vict0rsch) [reply omnigan/classifier.py]: I was giving you the benefit of the doubt ;)
